### PR TITLE
[NT 1991] UX Spinner while loading comments

### PR DESCRIPTION
**Pull request by moderateepheezy**
# 📲 What

Add a Loading indicator when loading more comments. 

# 🤔 Why

When user scrolls to the bottom we automatically load more comments, at the point of fetching more comments, we don't want users with slow network to think there are no more comments when we are just loading more of them.

# 🛠 How

1. Add a Custom TableView Footer with an activity indicator
2. Add the Custom TableView Footer when we subscribe to an event that triggers user is 3 comments aways to the end of the list.

# 👀 See

Trello, screenshots, external resources?

Loadmore

# ✅ Acceptance criteria

- [ ] An activity indicator shows when loading more comments and stops when new comments are fetched.

### CODE REVIEW

**Comment by nativeksr** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Kickstarter-iOS/Views/Controllers/CommentsViewController.swift
Comment (singhhari) [inline Kickstarter-iOS/Views/Controllers/CommentsViewController.swift:28]: Why is y 1?

File: Library/ViewModels/CommentsViewModel.swift
Comment (singhhari) [inline Library/ViewModels/CommentsViewModel.swift:44]: I think `shouldShowIndicator` is sufficient.

Comment (singhhari) [reply Library/ViewModels/CommentsViewModel.swift:44]: Also update the documentation to:
"Emits a Bool that determines if the activity indicator should render."

File: Library/ViewModels/CommentsViewModel.swift
Comment (singhhari) [inline Library/ViewModels/CommentsViewModel.swift:156]: if you're mapping `isCloseToBottom` to true every time, is it needed to combine here?

File: Library/ViewModels/CommentsViewModelTests.swift
Comment (singhhari) [inline Library/ViewModels/CommentsViewModelTests.swift:20]: Same renaming as above.

File: Library/ViewModels/CommentsViewModelTests.swift
Comment (singhhari) [inline Library/ViewModels/CommentsViewModelTests.swift:122]: If you are not changing the template, I would just delete the constant and supply an argument of `.template` wherever it's needed.

Comment (singhhari) [reply Library/ViewModels/CommentsViewModelTests.swift:122]: Check all instances of this below as well.

**Review by singhhari — CHANGES_REQUESTED** [review]

File: Library/ViewModels/CommentsViewModel.swift
Comment (moderateepheezy) [reply Library/ViewModels/CommentsViewModel.swift:156]: Yes, it's needed. `isCloseToBottom` ordinary returns void, so if it emits a value then I can map it as true. The condition is only satisfy if `isCloseToBottom` emits and `isCommentsLoading` emits value.

File: Library/ViewModels/CommentsViewModel.swift
Comment (moderateepheezy) [reply Library/ViewModels/CommentsViewModel.swift:44]: Isn't `shouldShowIndicator` a bit abstract since it doesn't say exactly what the indicator is about?

File: Kickstarter-iOS/Views/Controllers/CommentsViewController.swift
Comment (moderateepheezy) [reply Kickstarter-iOS/Views/Controllers/CommentsViewController.swift:28]: Haa, thanks I missed that.

File: Library/ViewModels/CommentsViewModelTests.swift
Comment (moderateepheezy) [reply Library/ViewModels/CommentsViewModelTests.swift:122]: Oh I didn't noticed it, I only copied out all the nested test. I'll remove the instances.

File: Kickstarter-iOS/Views/CommentTableViewFooter.swift
Comment (msadoon) [inline Kickstarter-iOS/Views/CommentTableViewFooter.swift:11]: Can just do `|> \.hidesWhenStopped .~ true`

File: Kickstarter-iOS/Views/Controllers/CommentsViewController.swift
Comment (msadoon) [inline Kickstarter-iOS/Views/Controllers/CommentsViewController.swift:28]: Can you use the `tableView` width instead?

File: Kickstarter-iOS/Views/CommentTableViewFooter.swift
Comment (msadoon) [inline Kickstarter-iOS/Views/CommentTableViewFooter.swift:15]: Not sure if this should be updated via a view model? cc @justinswart

File: Kickstarter.xcodeproj/project.pbxproj
Comment (msadoon) [inline Kickstarter.xcodeproj/project.pbxproj:635]: Might want to rename `CommentsTableViewFooter`

File: Library/ViewModels/CommentsViewModel.swift
Comment (singhhari) [reply Library/ViewModels/CommentsViewModel.swift]: The way I was looking at it was that there is no other "indicator" in the file but if you want to be slightly more descriptive you can do: `shouldShowActivityIndicator`

File: Library/ViewModels/CommentsViewModelTests.swift
Comment (msadoon) [inline Library/ViewModels/CommentsViewModelTests.swift:342]: May want to test `shouldShowIndicator` state here as well.

File: Library/ViewModels/CommentsViewModelTests.swift
Comment (msadoon) [inline Library/ViewModels/CommentsViewModelTests.swift:366]: test `shouldShowIndicator` here as well.

File: Library/ViewModels/CommentsViewModelTests.swift
Comment (msadoon) [inline Library/ViewModels/CommentsViewModelTests.swift:320]: If adding a test condition, would be good to update the name of the test.

**Review by msadoon — CHANGES_REQUESTED** [review]
Few changes, mostly around testing, otherwise looks great!
There was an issue with the scrolling indicator looking like it didn't hit the bottom of the scrollview when the loading indicator is shown:

Maybe also make the indicator larger and more centered?

File: Library/ViewModels/CommentsViewModel.swift
Comment (singhhari) [reply Library/ViewModels/CommentsViewModel.swift:156]: I guess what I'm suggesting is that you probably don't need to be calling mapConst(true) here and checking for it's value because it will always be true. Something like this will probably give you the same result:

    self.shouldShowIndicator = Signal
      .combineLatest(isCloseToBottom, self.isCommentsLoading)
      .map { _, isCommentsLoading in isCommentsLoading }
      
      maybe using `takeWhen(isCloseToBottom)` would give the same desired effect too but I'm not 100 percent certain.

File: Library/ViewModels/CommentsViewModelTests.swift
Comment (moderateepheezy) [reply Library/ViewModels/CommentsViewModelTests.swift:342]: There's a test for it, some few lines below.

File: Library/ViewModels/CommentsViewModel.swift
Comment (moderateepheezy) [reply Library/ViewModels/CommentsViewModel.swift]: I think we can leave it at `shouldShowIndicator `, we might not be able to use `shouldShowRefreshControl ` because it's not for refresh, but a load more.

File: Library/ViewModels/CommentsViewModelTests.swift
Comment (moderateepheezy) [reply Library/ViewModels/CommentsViewModelTests.swift:366]: As you've mentioned in one of the comments, rename would do well, testing `shouldShowIndicator` would return empty at those stages you mentioned.

File: Library/ViewModels/CommentsViewModel.swift
Comment (moderateepheezy) [reply Library/ViewModels/CommentsViewModel.swift]: @singhhari so I updated to this

```
self.shouldShowIndicator = Signal
  .combineLatest(isCloseToBottom, self.isCommentsLoading)
  .map { _, isCommentsLoading in isCommentsLoading }
```

But note that `isCloseToBottom.mapConst(true)` will only emit only when user scroll close to the bottom of the list. I have tested this with small and large comments.

Also `takeWhen(isCloseToBottom)` won't work in this case because you always need a `true` or `false` value.

File: Library/ViewModels/CommentsViewModelTests.swift
Comment (msadoon) [reply Library/ViewModels/CommentsViewModelTests.swift]: I guess the idea is to test the "before" and "after" states of this output, just like the other ones.

**Review by msadoon — APPROVED** [review]
looks good to me!

File: Library/ViewModels/CommentsViewModel.swift
Comment (singhhari) [reply Library/ViewModels/CommentsViewModel.swift]: Fair enough although you make a good point that it does not say what exactly the indicator is doing so something like `shouldShowLoadingIndicator` is half way between what you and I are suggesting.

File: Library/ViewModels/CommentsViewModel.swift
Comment (singhhari) [reply Library/ViewModels/CommentsViewModel.swift]: Yup sounds good. I just wanted to simplify it without compromising functionality.

File: Kickstarter-iOS/Views/CommentTableViewFooter.swift
Comment (singhhari) [reply Kickstarter-iOS/Views/CommentTableViewFooter.swift]: We almost always use a view model output to update a component of the UI. e.g `.hidden`

**Review by singhhari — CHANGES_REQUESTED** [review]
Coverage shouldn't be going down given the addition of new code. Should be good to go after code coverage.

**Comment by moderateepheezy** [comment]
> Coverage shouldn't be going down given the addition of new code. Should be good to go after code coverage.

I think this was related to the merge from @msadoon, the codecov was up by about 0.32%, I think it will go backup after I fix the merge conflicts and push again.

**Comment by msadoon** [comment]
> > Coverage shouldn't be going down given the addition of new code. Should be good to go after code coverage.
> 
> I think this was related to the merge from @msadoon last PR, the codecov was up by about 0.32%, I think it will go backup after I fix the merge conflicts and push again.

I think the last PR I merged had +0.01% code cov.

File: Kickstarter-iOS/Views/CommentTableViewFooterView.swift
Comment (singhhari) [inline Kickstarter-iOS/Views/CommentTableViewFooterView.swift:56]: There is a disconnect with the name of this output and the name of `shouldShowActivityIndicator`. We should make them more similar.

File: Kickstarter-iOS/Views/Controllers/CommentsViewController.swift
Comment (singhhari) [inline Kickstarter-iOS/Views/Controllers/CommentsViewController.swift:156]: there should be a capture list here.

File: Library/ViewModels/CommentTableViewFooterViewModelTests.swift
Comment (singhhari) [inline Library/ViewModels/CommentTableViewFooterViewModelTests.swift:24]: I would separate the testing of true and false into two separate test cases for readability.

File: Library/ViewModels/CommentsViewModel.swift
Comment (singhhari) [inline Library/ViewModels/CommentsViewModel.swift:50]: alphabetize

File: Library/ViewModels/CommentsViewModel.swift
Comment (singhhari) [inline Library/ViewModels/CommentsViewModel.swift:230]: same note as above.

File: Library/ViewModels/CommentsViewModelTests.swift
Comment (singhhari) [inline Library/ViewModels/CommentsViewModelTests.swift:21]: alphabetize

File: Library/ViewModels/CommentsViewModelTests.swift
Comment (singhhari) [inline Library/ViewModels/CommentsViewModelTests.swift:327]: the names of these functions are a bit much imo. We should simplify it something like:

`testLoadCommentsAndLoadingIndicator_WillDisplayRowAtEnd`

The exact name is open for discussion but we should either simplify the name or move some of the tests out into a separate method.

**Review by singhhari — CHANGES_REQUESTED** [review]
Left some comments but also want to verify the following:

The screenshot for `Screenshots/_64/Kickstarter_Framework_iOSTests.CommentsViewControllerTests/testView_WithFailedRemovedAndSuccessfulComments_ShouldDisplayAll_Comments___lang_de@2x.png` changed. Is that expected?

**Comment by moderateepheezy** [comment]
> WithFailedRemovedAndSuccess

This is probably due to a merge conflict I resolved, I think one of the screenshots was listed in the conflict, but can't remember which, but I recorded the tests again.

File: Library/ViewModels/CommentTableViewFooterViewModelTests.swift
Comment (singhhari) [inline Library/ViewModels/CommentTableViewFooterViewModelTests.swift:17]: I know this is annoyingly specific but you should be following the conventions we already use in the project:

`test{OUTPUTNAME}_{VariableX}_{true/false}`

so in the case of these two tests it will be something like `testShouldShowActivityIndicator_Configure_True`

there's no set way to do it but we do follow this throughout the project

**Review by singhhari — APPROVED** [review]

File: Library/ViewModels/CommentTableViewFooterViewModelTests.swift
Comment (moderateepheezy) [reply Library/ViewModels/CommentTableViewFooterViewModelTests.swift:17]: Oh, Alright then.

**Pull request merged**
